Write the Genesis block in the BlockStorage

AndrejMitrovic opened the pull request:
Fixes #286

This or https://github.com/bpfkorea/agora/pull/194 will need a rebase depending on which gets pulled first. So I'll mark this a blocked for now.

AndrejMitrovic committed:
Write the Genesis block in the BlockStorage

Move the loading code into a separate load() method,
and get rid of the empty() method.

Loading should ideally never be done in the constructor.

Fixes #286

AndrejMitrovic commented:
Fixed the integration tests.

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -188,21 +194,32 @@ public class BlockStorage : IBlockStorage
 
         if (!this.path.exists)
             mkdirRecurse(this.path);
```
I'd say this should be moved to `load` as well

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -962,14 +979,17 @@ public class MemBlockStorage : IBlockStorage
     ***************************************************************************/
 
-    public bool isEmpty ()
+    public bool load ()
```
It looks like the return value is only used in tests ? I don't think we need this information to be honest, it looks like a leaky abstraction...

AndrejMitrovic replied:
Like I mentioned before, I need to add the UTXO set from the genesis block in the UTXO PR.

I should only add the UTXOs if the genesis block was added for the very first time.

AndrejMitrovic commented:
Ok put this on hold, I've got a simpler idea I'll work on later.

AndrejMitrovic commented:
So, there is a better way to do what I wanted.

The ledger should check if the UTXO database is empty. If it's empty, it can walk through the ledger storage and add each block's UTXOs. Simple.

Unblocked.

AndrejMitrovic commented:
Of course, it may be more efficient to get the mempool from the network in some cases, but for starters we can do it this way.

AndrejMitrovic committed:
Add the .length() method in UTXOSet

If the UTXOSet is empty, it will need
to be rebuilt. Therefore we need to know
its length.

AndrejMitrovic committed:
Ensure UTXOSet is rebuilt if it's empty

Rebuild the UTXOSet from the blockchain
which was loaded from disk.

This will also allow moving the adding
of the genesis block back to the BlockStorage.

AndrejMitrovic commented:
Updated. There are no boolean return values anymore. It should be looked at commit by commit.

I still need to add one more test: if the UTXOSet was empty but the blockchain in the storage had multiple blocks, the UTXOSet should be properly re-built.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -162,6 +164,26 @@ public class Ledger
+        scope (failure) assert(0);
+
+        // add the new UTXOs
+        block.txs.each!(tx => this.utxo_set.updateUtxoCache(tx));
```
`UTXO`, will need a rename

Geod24 reviewed (approved):
LGTM, just waiting for the last test you mentioned.

AndrejMitrovic committed:
Rename updateUtxoCache() => updateUTXOCache()

Consistency with acronym naming.

AndrejMitrovic commented:
Updated with a test-case.
